README.md
# Detectr.js

[![Latest Version](https://img.shields.io/github/release/rogeriotaques/detectr.js.svg)](https://github.com/rogeriotaques/detectr.js/releases)

Empower web sites and apps design through clever CSS selectors.

This plugin is a very small javascript library which gives you the ability to
write clever and specific CSS rules to refine your website design for each
different browser, platform, device and operating system.

_It has a really small footprint and doesn't have any library dependencies such as: jQuery, Mootools, (...). So,
you can use it alone, or combined with those packages._

## Get started

### 1. Get the source code or install it via `npm`

Start by cloning this repository to your local machine, or just download the
source-code into any folder of your PC. As an open-source, the source-code is
available in its natural format and minified.

or simply run this command in your terminal

```sh
$ npm install -S detectr.js
```

### 2. Copy necessary files

Import or copy the `dist/detectr.js` (minified) file to any directory of your website
or app. For example:

```
/ (root)
|-- img/
|-- css/
|-- js/
|    |-- detectr.js
|-- index.html
```

### 3. Import the library

Import the library in your HTML files.

Since it is intended to help you create better designs by empower your control
on CSS rules according to used browsers, I strongly recommend you to import
Detectr.js within the HEAD tags.

E.g:

```
<html >
  <head >
    ...
    <script type="text/javascript" src="js/detectr.js" ></script>
  </head>
  <body >
    ...
  </body>
</html>

```

### 4. Done.

As simple as that. As soon as you complete the 3 steps above and refresh your
site in any browser, you will be able to see that many specific classes names
were added in the HTML tag of your page.

E.g:

```
<html class="webkit chrome chrome40 x32 mac osx osx10_10 js svg" >
  ...
</html>
```

## Available CSS classes

Detectr.js may add the following classes in the HTML tag:

### Browsers

| Class             | Remark                       |
| ----------------- | ---------------------------- |
| .edge             | Microsoft Edge               |
| .ie               | Internet Explorer            |
| .ie{version}      | E.g. ie10.                   |
| .webkit           | Browsers based on the webkit |
| .iron             | Iron Web Browser             |
| .chrome           | Google Chrome                |
| .chrome{version}  | E.g. chrome23                |
| .firefox          | Firefox Browser              |
| .firefox{version} | E.g. firefox32               |
| .opera            | Opera Browser                |
| .opera{version}   | E.g. opera16                 |
| .konqueror        | Konqueror Browser            |
| .safari           | Safari Browser               |
| .safari{version}  | E.g. safari12                |
| .gecko            | Browsers based on gecko      |
| .android-browser  | Android native browsers      |

### Operating Systems

| Class         | Remark                        |
| ------------- | ----------------------------- |
| .mac          | Apple Macintosh               |
| .osx{version} | E.g. osx10_10                 |
| .ios{version} | E.g. ios8                     |
| .windows      | Microsoft Windows             |
| .win2k        | Microsoft Windows 2000        |
| .win2k.sp1    | Microsoft Windows 2000 SP1    |
| .xp           | Microsoft Windows XP          |
| .vista        | Microsoft Windows Vista       |
| .win7         | Microsoft Windows 7           |
| .win8         | Microsoft Windows 8           |
| .win8_1       | Microsoft Windows 8.1         |
| .win10        | Microsoft Windows 10          |
| .nt           | Microsoft Windows NT          |
| .nt{version}  | E.g. nt6.1                    |
| .webtv        | Web TV                        |
| .freebsd      | Freebsd                       |
| .linux        | Linux                         |
| .android      | Android                       |
| .bb{version}  | E.g. bb10 (for Blackberry 10) |

### Platform

| Class | Remark           |
| ----- | ---------------- |
| .x32  | Platform 32 bits |
| .x64  | Platform 64 bits |
| .arm  | Platform ARM     |

### Devices

| Class       | Remark            |
| ----------- | ----------------- |
| .j2me       |                   |
| .iphone     | Apple iPhone      |
| .ipad       | Apple iPad        |
| .ipod       | Apple iPod        |
| .mobile     | Any mobile device |
| .blackberry | Blackberry        |

### Other

| Class      | Remark                       |
| ---------- | ---------------------------- |
| .touch     | Touch screen devices         |
| .js        | Javascript supported         |
| .svg       | SVG Images supported         |
| .no-svg    | SVG Images NOT supported     |
| .retina    | Devices with pixelratio > 1  |
| .no-retina | Devices with pixelration = 1 |
| .portrait  | Portrait orientation         |
| .landscape | Landscape orientation        |

## Inspiration

This project is based on Rafael Lima's library "css_browser_selector"
(http://rafael.adm.br/css_browser_selector/) which seems to be discontinued
since Nov 2nd, 2010.

## License

This projec is licensed under MIT License.

Copyright (c) 2015-2019 Rogério Taques

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Release notes

- v1.8.1 ~ v1.8.2 - Bundling with gulp and webpack.
- v1.8.0 - Add editorconfig, eslint and prevent wiping classes from HTML before injection.
- v1.7.0 - Fixed a small bug regarding global variable
- v1.6.0 - Add a listener to reanalyse page on resize event
- v1.4.0 ~ v1.5.0 - Bug fixes for detecting Android Stock Browser and some improvements
- v1.3.0 - Start detecting Android Stock Browser
- v1.2.0 - Start detecting when device runs iOS
- v1.1.0 - Added support for Microsoft Edge
- v1.0.0 - First release. Totally rewritten and new detectings added.


gulpfile.js
/*
 * Repotenshoku.com
 * © 2018, Sppoted, Inc.
 */

// ~~ IMPORTS ~~
const gulp = require('gulp');
const eslint = require('gulp-eslint');
const named = require('vinyl-named');
const webpackStream = require('webpack-stream');
const path = require('path');

// ~~ VARS ~~

const PATH = {
  DIST: './dist',
  SRC: './src'
};

// ~~ TASKS ~~

// Lint the JS code.
gulp.task(
  'lint',
  () => {
    gulp
      .src([`${PATH.SRC}/**/*.js`])
      .pipe(eslint())
      .pipe(eslint.format())
  }
);

// Transpile, parse and bundle javascript files
gulp.task('build', (cb) => {
  const webpackConfig = {
    mode: 'production',

    devtool: 'source-map',

    resolve: {
      modules: [path.resolve('node_modules'), path.resolve('src')]
    },

    module: {
      rules: [
        {
          test: /\.js$/,
          exclude: /(node_modules|bower_components)/,
          use: {
            loader: 'babel-loader',
            options: {
              presets: ['@babel/preset-env']
            }
          }
        }
      ]
    }
  };

  gulp
    .src([`${PATH.SRC}/detectr.js`])
    .pipe(named())
    .pipe(
      webpackStream(webpackConfig).on('error', () => { })
    )
    .pipe(gulp.dest(PATH.DIST));

  cb();
});

// Start development ...
gulp.task('default', gulp.series(['lint', 'build']));


src/detectr.js
/*!
 * Detectr.js
 * @author Rogerio Taques (rogerio.taques@gmail.com)
 * @see http://github.com/rogeriotaques/detectrjs
 *
 * This project is based on the Rafael Lima's work
 * which is called css_browser_selector and seems
 * to be discontinued. (http://rafael.adm.br/css_browser_selector/).
 */

const DetectrJs = (function detecterjs($) {
  const version = '1.8.1'

  /**
   * Whenever .trim() isn't supported, makes it be.
   */
  if (typeof String.prototype.trim !== 'function') {
    // eslint-disable-next-line no-extend-native
    String.prototype.trim = function trim() {
      return this.replace(/^\s+|\s+$/g, '')
    }
  }

  const doc = $.document
  const element = doc.documentElement

  const detectr = function detectr(userAgent) {
    const ua = userAgent.toLowerCase()
    const winWidth = $.outerWidth || element.clientWidth
    const winHeight = $.outerHeight || element.clientHeight
    const gecko = 'gecko'
    const webkit = 'webkit'
    const safari = 'safari'
    const opera = 'opera'
    const mobile = 'mobile'

    /**
     * Checks if given string is present on the userAgent.
     * @param  string  str
     * @return {Boolean}
     */
    const is = function is(str) {
      return ua.indexOf(str) > -1
    }

    /**
     * The core feature ...
     */
    let detect = function detect() {
      const rendered = []
      const { implementation } = doc
      const webkitVersion = /applewebkit\/(\d{1,})/.test(ua) ? RegExp.$1 : false

      let sysVersion = ''

      // *** Detecting browsers ***
      switch (true) {
        // internet explorer
        case (is('msie') && !is('opera') && !is('webtv')) || is('trident') || is('edge'):
          if (is('edge')) {
            sysVersion = /edge\/(\w+)/.test(ua) ? ' edge ie' + RegExp.$1 : ' ie11'
          } else if (is('msie 8.0') || is('trident/7.0')) {
            sysVersion = ' ie11'
          } else {
            sysVersion = /msie\s(\d+)/.test(ua) ? ' ie' + RegExp.$1 : ''
          }

          rendered.push('ie' + sysVersion)
          break

        // iron
        case is('iron/') || is('iron'):
          sysVersion = /iron\/(\d+)/.test(ua) ? ' iron' + RegExp.$1 : ''
          rendered.push(webkit + ' iron' + sysVersion)
          break

        // android
        case is('android') && is('u;') && (!is('chrome') || (is('chrome') && webkitVersion && webkitVersion <= 534)):
          // according to some researches android stock (native) browsers never went above applewebkit/534.x,
          // then, we can suppose user is using a native browser in android when the UA contains "android",
          // "mobile" and "U;" strings
          // @see: (http://stackoverflow.com/questions/14403766/how-to-detect-the-stock-android-browser)

          rendered.push('android-browser')
          break

        // google chrome
        case is('chrome/') || is('chrome'):
          sysVersion = /chrome\/(\d+)/.test(ua) ? ' chrome' + RegExp.$1 : ''
          rendered.push(webkit + ' chrome' + sysVersion)
          break

        // firefox
        case is('firefox/') || is('firefox'):
          sysVersion = /firefox\/(\d+)/.test(ua) ? ' firefox' + RegExp.$1 : ''
          rendered.push(gecko + ' firefox' + sysVersion)
          break

        // opera
        case is('opera/') || is('opera'):
          sysVersion = /version(\s|\/)(\d+)/.test(ua) || /opera(\s|\/)(\d+)/.test(ua) ? ' ' + opera + RegExp.$2 : ''
          rendered.push(opera + sysVersion)
          break

        // konqueror
        case is('konqueror'):
          rendered.push(mobile + ' konqueror')
          break

        // blackberry
        case is('blackberry') || is('bb'):
          rendered.push(mobile + ' blackberry')

          if (is('bb')) {
            sysVersion = /bb(\d{1,2})(;{0,1})/.test(ua) ? 'bb' + RegExp.$1 : ''
            rendered.push(sysVersion)
          }

          break

        // safari
        case is('safari/') || is('safari'):
          sysVersion = /version\/(\d+)/.test(ua) || /safari\/(\d+)/.test(ua) ? ' ' + safari + RegExp.$1 : ''
          rendered.push(webkit + ' ' + safari + sysVersion)
          break

        // applewebkit
        case is('applewebkit/') || is('applewebkit'):
          sysVersion = /applewebkit\/(\d+)/.test(ua) ? ' ' + webkit + RegExp.$1 : ''
          rendered.push(webkit + ' ' + sysVersion)
          break

        // gecko || mozilla
        case is('gecko') || is('mozilla/'):
          rendered.push(gecko)
          break

        default:
          break
      }

      // *** Detecting O.S ***
      switch (true) {
        // ios
        case is('iphone') || is('ios'):
          sysVersion = /iphone\sos\s(\d{1,2})/.test(ua) ? ' ios' + RegExp.$1 : ''

          // For some reason when it's iOS8, userAgent comes like OS 10_10
          // what returns a wrong version, then we need to match it against
          // another value
          if (sysVersion === ' ios10') {
            const vv = /(\d{1,2})/.test(sysVersion) ? RegExp.$1 : 0
            const vd = /\sversion\/(\d{1,2})/.test(ua) ? RegExp.$1 : ''

            if (parseInt(vv, 10) > parseInt(vd, 10)) {
              sysVersion = ' ios' + vd
            }
          }

          rendered.push('ios' + sysVersion)
          break

        // macintosh
        case is('mac') || is('macintosh') || is('darwin'):
          sysVersion = /mac\sos\sx\s(\d{1,2}_\d{1,2})/.test(ua) ? ' osx' + RegExp.$1 : ''
          rendered.push('mac' + sysVersion)
          break

        // windows
        case is('windows') || is('win'):
          sysVersion = /windows\s(nt\s{0,1})(\d{1,2}\.\d)/.test(ua) ? '' + RegExp.$2 : ''

          // defining windows version
          switch (sysVersion) {
            case '5.0':
              sysVersion = ' win2k'
              break
            case '5.01':
              sysVersion = ' win2k sp1'
              break
            case '5.1':
            case '5.2':
              sysVersion = ' xp'
              break
            case '6.0':
              sysVersion = ' vista'
              break
            case '6.1':
              sysVersion = ' win7'
              break
            case '6.2':
              sysVersion = ' win8'
              break
            case '6.3':
              sysVersion = ' win8_1'
              break
            case '6.4':
              sysVersion = ' win10'
              break
            default:
              sysVersion = ' nt nt' + sysVersion
          }

          rendered.push('windows' + sysVersion)
          break

        // webtv
        case is('webtv'):
          rendered.push('webtv')
          break

        // freebsd
        case is('freebsd'):
          rendered.push('freebsd')
          break

        // android
        case is('android') || (is('linux') && is('mobile')):
          rendered.push('android')
          break

        // linux
        case is('linux') || is('x11'):
          rendered.push('linux')
          break

        default:
          break
      }

      // *** Detecting platform ***
      switch (true) {
        // 64 bits
        case is('wow64') || is('x64'):
          rendered.push('x64')
          break

        // arm
        case is('arm'):
          rendered.push('arm')
          break

        // 32 bits
        default:
          rendered.push('x32')
      }

      // *** Detecting devices ***
      switch (true) {
        case is('j2me'):
          rendered.push(mobile + ' j2me')
          break

        case /(iphone|ipad|ipod)/.test(ua):
          rendered.push(mobile + ' ' + RegExp.$1)
          break

        case is('mobile'):
          rendered.push(mobile)
          break

        default:
          break
      }

      // *** Detecting touchable devices ***
      if (/touch/.test(ua)) {
        rendered.push('touch')
      }

      // *** Assume that it supports javascript by default ***
      rendered.push('js')

      // *** Detect if SVG images are supported ***
      rendered.push(
        implementation !== undefined &&
          typeof implementation.hasFeature === 'function' &&
          implementation.hasFeature('http://www.w3.org/TR/SVG11/feature#Image', '1.1')
          ? 'svg'
          : 'no-svg'
      )

      // *** Detect retina display ***
      rendered.push($.devicePixelRatio !== undefined && $.devicePixelRatio > 1 ? 'retina' : 'no-retina')

      // *** Detecting orientation ***
      rendered.push(winWidth < winHeight ? 'portrait' : 'landscape')

      return rendered
    }

    // retrieve current classes attached
    let currentClassNames = doc.documentElement.className.split(' ')

    // convert 'detect' from function to array
    // and avoid unnecessary processing
    detect = detect()

    // concat all detected classes to the existing ones and make sure they are unique
    // this prevent wiping pre-existing classes attached by different processes.
    currentClassNames = currentClassNames.concat(detect)
    currentClassNames = currentClassNames.filter((v, i) => currentClassNames.indexOf(v) === i)

    // inject the new classes set in the HTML tag.
    element.className = currentClassNames
      .join(' ')
      .trim()

    // return what was detected
    return {
      detected: detect.join(' ').trim(),
      version
    }
  }

  // execute and exposes detectr.js to the browser
  // eslint-disable-next-line
  $.detectr = detectr($.navigator.userAgent)

  /**
   * The listener engine for resize event ...
   */
  const resizing = function resizing() {
    $.detectr = detectr($.navigator.userAgent) // eslint-disable-line
  }

  // add an event listener for window resize
  // which will asure that references will be
  // updated in case of browser resizing
  if ($.attachEvent) {
    $.attachEvent('onresize', resizing)
  } else if ($.addEventListener) {
    $.addEventListener('resize', resizing, true)
  }
})

DetectrJs(window)

export default DetectrJs
